### PR TITLE
initial memberlist.Transport implementation allowing us to use an upgrader for ListenTCP/ListenUDP

### PR DESCRIPTION
**Pull request by amery**

### CODE REVIEW

File: transport/tcp.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+import (
+	"context"
+	"net"
+	"time"
+
+	"github.com/hashicorp/memberlist"
+)
+
+// DialAddressTimeout is used by memberlist to establish a TCP connection to a particular node
+func (t *Transport) DialAddressTimeout(addr memberlist.Address, timeout time.Duration) (net.Conn, error) {
```
Comment (karasz) [inline transport/tcp.go:12]: You are not using ```t```

```go
func (_ *Transport) DialAddressTimeout(addr memberlist.Address, timeout time.Duration) (net.Conn, error) {
...
```

File: cluster.go [hunk 2]
```diff
@@ -2,6 +2,7 @@ package gossipcache
 
 import (
```
Comment (karasz) [inline cluster.go:3]: you have a whitespace issue on https://github.com/darvaza-proxy/gossipcache/blob/35357924f5f69c94d5a08c21c811343ee874b6f6/cluster.go#L293

File: transport/tcp.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+			case <-ctx.Done():
+				// sorry pal, we are cancelled
+				defer conn.Close()
+
+				t.error(ctx.Err()).
+					WithField(ListenerAddrLabel, ln.Addr()).
+					WithField(RemoteAddrLabel, conn.RemoteAddr()).
+					Print("Connection Terminated")
+
+				break
```
Comment (karasz) [inline transport/tcp.go:67]: does this ```break``` do really what you want? Didn't you mean:
```go
mainfor:for {
....
break mainfor
}
```

File: transport/udp.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+			case t.packetCh <- msg:
+				// continue
+			case <-ctx.Done():
+				// cancelled
+				t.error(ctx.Err()).
+					WithField(ListenerAddrLabel, ln.LocalAddr()).
+					WithField(RemoteAddrLabel, addr).
+					WithField(PacketSizeLabel, n).
+					Print("UDP packet discarded")
+				break
```
Comment (karasz) [inline transport/udp.go:108]: see the comment in tcp.co

File: transport/tcp.go (hunk 1 above)
Comment (karasz) [inline transport/tcp.go:12]: you may want:
```go
func (_ *Transport) DialAddressTimeout(addr memberlist.Address, timeout time.Duration) (net.Conn, error) {
...
}
```
as you are not using ```t``` in the function

**Review by karasz — CHANGES_REQUESTED** [review]
Other than what was commented this LGTM

File: transport/tcp.go (hunk 3 above)
Comment (amery) [reply transport/tcp.go:67]: good catch, but what about using `return` instead? @karasz

File: transport/tcp.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+				WithField(ListenerAddrLabel, ln.Addr()).
+				WithField(RemoteAddrLabel, conn.RemoteAddr()).
+				Print("Connected")
+
+			select {
+			case t.streamCh <- conn:
+				// continue
+			case <-ctx.Done():
+				// sorry pal, we are cancelled
+				defer conn.Close()
```
[the remark this answers is not in the document]
Comment (amery) [reply transport/tcp.go:60]: combined with `return` becomes more explicit that it only happens once

File: transport/tcp.go (hunk 1 above)
Comment (karasz) [reply transport/tcp.go:12]: either works. possibly removing entirely is cleaner indeed.

**Comment by amery** [comment]
amended. also updated it to use the new slog/handlers/discard instead of slog/noop @karasz

File: transport/tcp.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (karasz) [reply transport/tcp.go:60]: This is still not fixed, line 60 is unchanged

[the remark this answers is not in the document]
Comment (amery) [reply transport/tcp.go:60]: It's not in a loop because there is a return right after

**Review by karasz — APPROVED** [review]
LGTM now

**Pull request merged**